pkg/controller/accept/accept_controller.go

```diff
@@ -0,0 +1,474 @@
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("Dropping cluster out of the queue: %v", err)
+	c.queue.Forget(key)
+}
+
+type jobFactory func(string) (*v1batch.Job, *kapi.ConfigMap, error)
+
+func (f jobFactory) BuildJob(name string) (*v1batch.Job, *kapi.ConfigMap, error) {
+	return f(name)
+}
+
+type jobOwnerControl struct {
+	controller *Controller
+}
+
+func (c *jobOwnerControl) GetOwnerKey(owner metav1.Object) (string, error) {
+	return controller.KeyFunc(owner)
+}
+
+func (c *jobOwnerControl) GetOwner(namespace string, name string) (metav1.Object, error) {
+	return c.controller.clustersLister.Clusters(namespace).Get(name)
+}
+
+func (c *jobOwnerControl) OnOwnedJobEvent(owner metav1.Object) {
+	cluster, ok := owner.(*clusteroperator.Cluster)
+	if !ok {
+		c.controller.logger.WithFields(log.Fields{"owner": owner.GetName(), "namespace": owner.GetNamespace()}).
+			Errorf("attempt to enqueue owner that is not a cluster")
+		return
+	}
+	c.controller.enqueueCluster(cluster)
+}
+
+type jobSyncStrategy struct {
+	controller *Controller
+}
+
+func (s *jobSyncStrategy) GetOwner(key string) (metav1.Object, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if len(namespace) == 0 || len(name) == 0 {
+		return nil, fmt.Errorf("invalid key %q: either namespace or name is missing", key)
+	}
+	return s.controller.clustersLister.Clusters(namespace).Get(name)
+}
+
+func (s *jobSyncStrategy) DoesOwnerNeedProcessing(owner metav1.Object) bool {
+	cluster, ok := owner.(*clusteroperator.Cluster)
+	if !ok {
+		s.controller.logger.Warn("could not convert owner from JobSync into a cluster: %#v", owner)
+		return false
+	}
+	if cluster.Status.Accepted && cluster.Status.AcceptedJobGeneration == cluster.Generation {
+		s.controller.logger.Debugf("cluster %s/%s is already accepted for its current generation, it will not be processed", cluster.Namespace, cluster.Name)
+		return false
+	}
+
+	machineSets, err := controller.MachineSetsForCluster(cluster, s.controller.machineSetsLister)
+	if err != nil {
+		s.controller.logger.Errorf("error retrieving machine sets for cluster %s/%s: %v", cluster.Namespace, cluster.Name, err)
+		utilruntime.HandleError(fmt.Errorf("cannot retrieve machinesets for cluster %s/%s: %v", cluster.Namespace, cluster.Name, err))
+		return false
+	}
+
+	// Only process if all machinesets have been provisioned
+	totalCount := len(machineSets)
+	provisionedCount := 0
+	installedCount := 0
+	for _, ms := range machineSets {
+		if ms.Status.Provisioned {
+			provisionedCount++
+		}
+		if ms.Spec.NodeType == clusteroperator.NodeTypeMaster && ms.Status.Installed {
+			installedCount++
+		}
+	}
+
+	// With at least one machine set, all machineSets in the cluster provisioned and a
+	// master that is installed, we are ready to accept nodes
+	return totalCount > 0 && provisionedCount == totalCount && installedCount > 0
```

Comment:
This check is a bit more complex than it needs to be. Why don't we simply return false if we find a machine set that is not provisioned or a master machine set that is not installed. In other words, replace `provisionedCount++` and `installedCount++` with `return false`?